File: technole-raee-project/technole-raee/backend/python/routers/clients.py
"""
routers/clients.py
"""
from fastapi     import APIRouter, Depends, HTTPException
from typing      import List
from models.schemas import Client, ClientCreate, ClientUpdate
from db.database    import get_db
from middleware.auth import require_auth

router = APIRouter()

@router.get("/", response_model=List[Client])
async def get_clients(db=Depends(get_db), _=Depends(require_auth)):
    rows = await db.fetch("SELECT * FROM clients ORDER BY name")
    return [dict(r) for r in rows]

@router.post("/", response_model=Client, status_code=201)
async def create_client(body: ClientCreate, db=Depends(get_db), _=Depends(require_auth)):
    row = await db.fetchrow(
        "INSERT INTO clients (name,contact,email,phone,nif) VALUES ($1,$2,$3,$4,$5) RETURNING *",
        body.name, body.contact, body.email, body.phone, body.nif
    )
    return dict(row)

@router.put("/{client_id}", response_model=Client)
async def update_client(client_id: int, body: ClientUpdate, db=Depends(get_db), _=Depends(require_auth)):
    row = await db.fetchrow(
        "UPDATE clients SET name=$1,contact=$2,email=$3,phone=$4,nif=$5 WHERE id=$6 RETURNING *",
        body.name, body.contact, body.email, body.phone, body.nif, client_id
    )
    if not row: raise HTTPException(404, "Client not found")
    return dict(row)

@router.delete("/{client_id}", status_code=204)
async def delete_client(client_id: int, db=Depends(get_db), _=Depends(require_auth)):
    await db.execute("DELETE FROM clients WHERE id=$1", client_id)

File: technole-raee-project/technole-raee/backend/node/routes/clients.js
/**
 * routes/clients.js
 * CRUD completo de clientes.
 */

import { Router } from "express";
import { pool }   from "../db/pool.js";
import { auth }   from "../middleware/auth.js";

export const clientsRouter = Router();

// GET /api/clients
clientsRouter.get("/", auth, async (_req, res, next) => {
  try {
    const { rows } = await pool.query(
      "SELECT * FROM clients ORDER BY name"
    );
    res.json(rows);
  } catch (e) { next(e); }
});

// POST /api/clients
clientsRouter.post("/", auth, async (req, res, next) => {
  const { name, contact, email, phone, nif } = req.body;
  try {
    const { rows } = await pool.query(
      "INSERT INTO clients (name, contact, email, phone, nif) VALUES ($1,$2,$3,$4,$5) RETURNING *",
      [name, contact, email, phone, nif]
    );
    res.status(201).json(rows[0]);
  } catch (e) { next(e); }
});

// PUT /api/clients/:id
clientsRouter.put("/:id", auth, async (req, res, next) => {
  const { name, contact, email, phone, nif } = req.body;
  try {
    const { rows } = await pool.query(
      "UPDATE clients SET name=$1, contact=$2, email=$3, phone=$4, nif=$5 WHERE id=$6 RETURNING *",
      [name, contact, email, phone, nif, req.params.id]
    );
    if (!rows.length) return res.status(404).json({ error: "Not found" });
    res.json(rows[0]);
  } catch (e) { next(e); }
});

// DELETE /api/clients/:id
clientsRouter.delete("/:id", auth, async (req, res, next) => {
  try {
    await pool.query("DELETE FROM clients WHERE id=$1", [req.params.id]);
    res.status(204).end();
  } catch (e) { next(e); }
});
